sql: migration and mixed cluster version roachtest for grant option

jackcwu opened the pull request:
The roachtest ensures that the validation check for grant options that is performed on a grant/revoke only runs when all nodes on a cluster are upgraded to 22.1. The migration scans through existing users and sets their grant option bits equal to their privilege bits if they have GRANT or ALL privileges (otherwise, it would have been impossible for them to grant anyways, so we do not do anything).

jackcwu committed:
sql: migration and mixed cluster version roachtest for grant option

The roachtest ensures that the validation check for
grant options that is performed on a grant/revoke only runs when all nodes on
a cluster are upgraded to 22.1. The migration scans through existing users
and sets their grant option bits equal to their privilege bits if they
have GRANT or ALL privileges (otherwise, it would have been impossible
for them to grant anyways, so we do not do anything).

Release note: None

RichardJCai commented on pkg/sql/catalog/descpb/privilege.go:
```diff
@@ -41,6 +41,17 @@ const (
+	// Version22_1 corresponds to desciptors created in 22.1 and onwards.
+	// These descriptors should have grant options for all their privileges
+	// if they have the GRANT or ALL were created prior to 22.1.
+	Version22_1
```
I don't think we need to introduce two versions here

I think we only need `GrantOptionVersion` such that on `GrantOptionVersion`, we're guaranteed that the grant option bits are set (ie the migration has run or that the `PrivilegeDesc` was created on a version where grant option is supported)

RichardJCai commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -80,6 +80,13 @@ type DescriptorBuilder interface {
+	// above. The difference is that these changes will only every be run during the
+	// migration and will stop once all nodes are upgraded and waitForUpgradeStep(c.All())
+	// is called.
+	RunMigrationOnlyChanges(ctx context.Context, dg DescGetter) error
```
I'd rather call this `MaybeAddWithGrantOptions` and note that we only want it to be run during the migration you're adding.

RichardJCai commented on pkg/sql/catalog/catprivilege/fix.go:
```diff
@@ -138,3 +138,58 @@ func MaybeFixPrivileges(
+func MaybeAddGrantOptions(ptr **descpb.PrivilegeDescriptor) bool {
+	p := *ptr
+	changed := false
+	if p.Version == descpb.GrantOptionVersion { // do not apply changes
```
This should be if `p.Version >= descpb.GrantOptionVersion`

RichardJCai commented on pkg/sql/catalog/catprivilege/fix.go:
```diff
@@ -138,3 +138,58 @@ func MaybeFixPrivileges(
+	//	p.SetVersion(descpb.Version21_2)
+	//	changed = true
+	//}
+	p.SetVersion(descpb.Version22_1)
```
Set to `GrantOptionVersion`

RichardJCai commented on pkg/migration/migrations/grant_option_migration.go:
```diff
@@ -0,0 +1,142 @@
+		return nil
+	}
+
+	return addGrantOptionMigration(ctx, rows, addGrantOptionFunc, 1<<19)
```
Add comment `1<<19 /* 512 KiB batch size */`

RichardJCai reviewed (commented):
Adding my high level comments first

jackcwu reviewed (commented):
Thank you, I made the fixes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-:-MpqrpBd9sbJKVj-ceWf:b-j2h19k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r2-67:-MpqfXI29xXPHSlcCzL3:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/migration/migrations/grant_option_migration.go#L67)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Add comment `1<<19 /* 512 KiB batch size */`
</blockquote></details>

added
___
*[pkg/sql/catalog/descriptor.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3h-r2-88:-MpqhhBJ6ljjB7V8gId8:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/descriptor.go#L88)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I'd rather call this `MaybeAddWithGrantOptions` and note that we only want it to be run during the migration you're adding.
</blockquote></details>

fixed
___
*[pkg/sql/catalog/catprivilege/fix.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3f-r2-148:-MpqjSR_2t7vVOTizf6M:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/catprivilege/fix.go#L148)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This should be if `p.Version >= descpb.GrantOptionVersion`
</blockquote></details>

fixed
___
*[pkg/sql/catalog/catprivilege/fix.go, line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3f-r2-191:-MpqjNm232FlEEgJllCq:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/catprivilege/fix.go#L191)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Set to `GrantOptionVersion`
</blockquote></details>

fixed
___
*[pkg/sql/catalog/descpb/privilege.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3Nk0-Tc7zYCyJWl-r2-54:-Mpqj5Gb07vflCliiBQk:b-tsmctb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/descpb/privilege.go#L54)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I don't think we need to introduce two versions here

I think we only need `GrantOptionVersion` such that on `GrantOptionVersion`, we're guaranteed that the grant option bits are set (ie the migration has run or that the `PrivilegeDesc` was created on a version where grant option is supported)
</blockquote></details>

my bad, this was something I was testing out when I was previously trying to incorporate these changes into `runPostDeserializationChanges` and forgot to remove. Fixed


<!-- Sent from Reviewable.io -->

RichardJCai commented on pkg/cmd/roachtest/tests/privilege_version_upgrade.go:
```diff
@@ -47,6 +47,11 @@ func runPrivilegeVersionUpgrade(
 	}
 
 	const currentVersion = ""
+	fmt.Println("<------------------------------------------------>")
```
Remove this please.

RichardJCai commented on pkg/cmd/roachtest/tests/validate_grant_option.go:
```diff
@@ -0,0 +1,134 @@
+
+// execSQLUser changes the current user from "root" to the one that is input and executes the
+// SQL statement as that user.
+func execSQLUser(sqlStatement string, user string, errorExpected bool, node int) versionStep {
```
Rename to `execAsUser` and update comment to say it executes the statement as the specified user.

RichardJCai commented on pkg/cmd/roachtest/tests/validate_grant_option.go:
```diff
@@ -0,0 +1,134 @@
+		t.Fatal(err)
+	}
+
+	u := newVersionUpgradeTest(c,
```
Can you add some comments here to make it easier to follow about what we're expecting at each "stage", ie what do we expect to happen before the migration runs, what do we expect to happen after the migration runs.

RichardJCai commented on pkg/migration/migrations/grant_option_migration.go:
```diff
@@ -0,0 +1,139 @@
+)
+
+// grantOptionMigration iterates through every descriptor and sets a user's grant option bits
+// equal to its privilege bits if it holds the "GRANT" privilege
```
nit: period

RichardJCai commented on pkg/migration/migrations/grant_option_migration.go:
```diff
@@ -0,0 +1,139 @@
+				return errors.Newf("unable to find descriptor for id %d", id)
+			}
+
+			err := b.MaybeAddGrantOptions(ctx, nil /* DescGetter */) // RunPostDeserializationChanges previously
```
Remove comment here

RichardJCai commented on pkg/sql/catalog/schemadesc/schema_desc_builder.go:
```diff
@@ -63,6 +63,17 @@ func (sdb *schemaDescriptorBuilder) RunPostDeserializationChanges(
+}
+
+// MaybeAddGrantOptions implements the catalog.DescriptorBuilder
+// interface
```
nit: period

RichardJCai commented on pkg/sql/catalog/tabledesc/table_desc_builder.go:
```diff
@@ -104,6 +104,20 @@ func (tdb *tableDescriptorBuilder) RunPostDeserializationChanges(
+	var err error
+	tdb.maybeModified = protoutil.Clone(tdb.original).(*descpb.TableDescriptor)
+	tdb.changes, err = maybeFillInDescriptor(ctx, dg, tdb.maybeModified, tdb.skipFKsWithNoMatchingTable)
+	tdb.changes, err = maybeAddGrantOptionsToDescriptor(tdb.maybeModified)
```
Do we need this helper function instead of doing what the helper does directly?

RichardJCai commented on pkg/sql/catalog/catprivilege/fix.go:
```diff
@@ -138,3 +138,54 @@ func MaybeFixPrivileges(
+
+// MaybeUpdateGrantOptions iterates over the users of the descriptor and checks
+// if they have the GRANT privilege - if so, then set the user's grant option
+// bits equal to the privilege bits
```
nit: period.

RichardJCai commented on pkg/sql/catalog/catprivilege/fix.go:
```diff
@@ -138,3 +138,54 @@ func MaybeFixPrivileges(
+func MaybeUpdateGrantOptions(ptr **descpb.PrivilegeDescriptor) bool {
+	p := *ptr
+	changed := false
+	if p.Version >= descpb.GrantOptionVersion { // do not apply changes
```
nit: period and specify why we do not need to apply changes (on >= `GrantOptionVersion` means grant bits are set)

RichardJCai commented:
cc @cockroachdb/sql-schema @ajwerner could y'all PTAL at the 2nd commit as well?

jackcwu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/cmd/roachtest/tests/privilege_version_upgrade.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3d-r3-50:-Mptcq8aB_5CfX8YTlj4:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/cmd/roachtest/tests/privilege_version_upgrade.go#L50)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Remove this please.
</blockquote></details>

removed
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3l-r3-50:-MptdJ_MBWHehrrC6E_0:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/cmd/roachtest/tests/validate_grant_option.go#L50)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Rename to `execAsUser` and update comment to say it executes the statement as the specified user.
</blockquote></details>

fixed
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3l-r3-72:-MptkT3iEOIcN0VYzWFy:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/cmd/roachtest/tests/validate_grant_option.go#L72)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Can you add some comments here to make it easier to follow about what we're expecting at each "stage", ie what do we expect to happen before the migration runs, what do we expect to happen after the migration runs.
</blockquote></details>

added
___
*[pkg/migration/migrations/grant_option_migration.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r3-30:-MptdQkY3yr451QtVxVy:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/migration/migrations/grant_option_migration.go#L30)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: period
</blockquote></details>

added
___
*[pkg/migration/migrations/grant_option_migration.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r3-50:-MptdWq55vf2oOq6ao6s:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/migration/migrations/grant_option_migration.go#L50)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Remove comment here
</blockquote></details>

removed
___
*[pkg/sql/catalog/catprivilege/fix.go, line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3f-r3-144:-Mptdf1b5jH3440_VKc3:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/sql/catalog/catprivilege/fix.go#L144)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

added
___
*[pkg/sql/catalog/catprivilege/fix.go, line 148 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3f-r3-148:-Mpte_hV0ZUiEC2MfSu8:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/sql/catalog/catprivilege/fix.go#L148)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: period and specify why we do not need to apply changes (on >= `GrantOptionVersion` means grant bits are set)
</blockquote></details>

added
___
*[pkg/sql/catalog/schemadesc/schema_desc_builder.go, line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3i-r3-71:-Mpte4mf1Cz9Ce8CdsPZ:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/sql/catalog/schemadesc/schema_desc_builder.go#L71)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: period
</blockquote></details>

added
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3j-r3-119:-MpvxilEAbaaqIU7sAG9:bstwn30) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/sql/catalog/tabledesc/table_desc_builder.go#L119)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Do we need this helper function instead of doing what the helper does directly?
</blockquote></details>

changed to call that in the function directly.

I realized that I accidentally left in `tdb.changes, err = maybeFillInDescriptor(ctx, dg, tdb.maybeModified, tdb.skipFKsWithNoMatchingTable)`, which is something called in `runPostDesserializationChanges`. I believe this means that the error can always be nil for this function.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jackcwu, @RichardJCai, and @stevendanna)

___
*[-- commits, line 4 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUoOvH6crbyDXra3og):*
nit: add newlines to the commit message. each line should be 80 characters
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUtMhrDVBokluPdh7V:-MqUtMhs60x_HrtOGbji:bykd11j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/cmd/roachtest/tests/validate_grant_option.go#L57)):*
> ```Go
> 			require.NoError(t, err)
> 		} else {
> 			require.Error(t, err)
> ```

this line should also assert that the error is what we expect it to be. maybe instead of `errorExpected` the param could be `expectedErrText` (and if it's an empty string, then keep the `require.NoError()`)
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 143 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUuHkh6SVqbe1B4-sR:-MqUuHkh6SVqbe1B4-sS:b6u72ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/cmd/roachtest/tests/validate_grant_option.go#L143)):*
> ```Go
> 		execSQL("CREATE USER foo5;", false, 2),
> 		execSQL("CREATE TABLE t2();", false, 3),
> 		execSQL("GRANT ALL PRIVILEGES ON TABLE t2 TO foo5;", false, 1),
> ```

can there be one more check here to make sure that `GRANT GRANT ...` is turned into `WITH GRANT OPTION` after the upgrade?

and maybe also another one to make sure `GRANT ... WITH GRANT OPTION` works as expected
___
*[pkg/migration/migrations/grant_option_migration.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r3-50:-MqUudNG6yHRE11JTNOL:b-4d2er3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/migration/migrations/grant_option_migration.go#L50)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

removed
</blockquote></details>

nit: remove commented code
___
*[pkg/migration/migrations/grant_option_migration.go, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUugFh5ahj05qsZzGS:-MqUugFh5ahj05qsZzGT:b1vsvns) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/migration/migrations/grant_option_migration.go#L36)):*
> ```Go
> 	query := `SELECT id, descriptor, crdb_internal_mvcc_timestamp FROM system.descriptor ORDER BY ID ASC`
> 	rows, err := d.InternalExecutor.QueryIterator(
> 		ctx, "update-grant-options", nil /* txn */, query,
> ```

nit: rename `update-grant-options` to `retrieve-grant-options`
___
*[pkg/sql/catalog/descriptor.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3h-r2-88:-MqUvVDJ9mbMnQv1Z-3q:b8arod9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/descriptor.go#L88)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

fixed
</blockquote></details>

nit: still needs to be renamed to `MaybeAddWithGrantOptions`
___
*[pkg/sql/catalog/catprivilege/fix.go, line 162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUwNe_4UgovTyK1nRk:-MqUwNe_4UgovTyK1nRl:b105goe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/sql/catalog/catprivilege/fix.go#L162)):*
> ```Go
> 
> 	// give ALL to the grant option bits for the root and admin users since they are superusers.
> 	superuserAllGrantOptions(security.RootUserName())
> ```

we don't really need this step, right? since root/admin skips privilege checks?


<!-- Sent from Reviewable.io -->

jackcwu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @stevendanna)

___
*[-- commits, line 4 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUoOvH6crbyDXra3og):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add newlines to the commit message. each line should be 80 characters
</blockquote></details>

fixed
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUtMhrDVBokluPdh7V:-MqoagTc8nvpGzVEAwAu:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/cmd/roachtest/tests/validate_grant_option.go#L57)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this line should also assert that the error is what we expect it to be. maybe instead of `errorExpected` the param could be `expectedErrText` (and if it's an empty string, then keep the `require.NoError()`)
</blockquote></details>

added
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 143 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUuHkh6SVqbe1B4-sR:-Mqp0QL_879sSOyC--fG:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/cmd/roachtest/tests/validate_grant_option.go#L143)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can there be one more check here to make sure that `GRANT GRANT ...` is turned into `WITH GRANT OPTION` after the upgrade?

and maybe also another one to make sure `GRANT ... WITH GRANT OPTION` works as expected
</blockquote></details>

added
___
*[pkg/migration/migrations/grant_option_migration.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r3-50:-MqmepeVCXCzBpvcLk32:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/migration/migrations/grant_option_migration.go#L50)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: remove commented code
</blockquote></details>

removed
___
*[pkg/migration/migrations/grant_option_migration.go, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUugFh5ahj05qsZzGS:-MqlKwa60OTRf9gQeayb:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/migration/migrations/grant_option_migration.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: rename `update-grant-options` to `retrieve-grant-options`
</blockquote></details>

fixed
___
*[pkg/sql/catalog/descriptor.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3h-r2-88:-MqmerBQCFpVgS0kRKTK:bq148w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e352827aa23528dc4ed4693df1a6b29ea7d460c9/pkg/sql/catalog/descriptor.go#L88)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: still needs to be renamed to `MaybeAddWithGrantOptions`
</blockquote></details>

changing this to `runMigrationOnlyChanges` instead as a generic method for that behavior
___
*[pkg/sql/catalog/catprivilege/fix.go, line 162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MqUwNe_4UgovTyK1nRk:-MqlMu5W5p_hqQYpIFUq:bi9lr5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2136cc0ea1ad6cc25dc8e0b61faab282c073022f/pkg/sql/catalog/catprivilege/fix.go#L162)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we don't really need this step, right? since root/admin skips privilege checks?
</blockquote></details>

true, removed


<!-- Sent from Reviewable.io -->

RichardJCai commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -64,6 +64,10 @@ const (
 
+// MigrationOnlyFunc represents the work that is applied when
+// RunMigrationOnlyChanges() is called
+type MigrationOnlyFunc = func(b DescriptorBuilder)
```
I'd rather call this `MigrationWorkFunc`

RichardJCai commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -80,6 +84,13 @@ type DescriptorBuilder interface {
+	// above. The difference is that these changes will only ever be run during the
+	// migration in grant_option_migration.go and will stop once all nodes are upgraded
+	// and waitForUpgradeStep(c.All()) is called.
+	RunMigrationOnlyChanges(workFunc ...MigrationOnlyFunc) error
```
Prefer `RunMigrationFuncs` or `RunMigrationChanges`

RichardJCai commented on pkg/sql/catalog/dbdesc/database_desc_builder.go:
```diff
@@ -73,6 +73,18 @@ func (ddb *databaseDescriptorBuilder) RunPostDeserializationChanges(
+// RunMigrationOnlyChanges implements the catalog.DescriptorBuilder
+// interface.
+func (ddb *databaseDescriptorBuilder) RunMigrationOnlyChanges(
+	workFunc ...catalog.MigrationOnlyFunc,
```
s/workFunc/workFuncs

rafiss reviewed (approved):
lgtm! just minor nits. let's get sql-schema to sign off for awareness

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-:-MrTSM0WCLQgqgisvtvu:b-nbn1q3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jackcwu, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 45 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRWMn7c7_gzbYe2Yd:-MrTRWMoB87UNaa_mgef:b1c38r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L45)):*
> ```Go
> 		} else {
> 			require.Error(t, err)
> 			if err.Error() != expectedErrText {
> ```

you can use `require.EqualError` instead of this manual check
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 63 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRaYZCG2dsqrOGwBR:-MrTRaYZCG2dsqrOGwBS:bjanwvq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L63)):*
> ```Go
> 		} else {
> 			require.Error(t, err)
> 			if err.Error() != expectedErrText {
> ```

nit: use `require.EqualError`
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 143 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRfM92j62FjaEpg-Z:-MrTRfM92j62FjaEpg-_:bmdj7z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L143)):*
> ```Go
> 		// option bits set equal to their privilege bits if they hold "GRANT" or "ALL" privileges. Grant options will
> 		// now be enforced (an error will be returned if a user tries to grant a privilege they do not hold grant
> 		// options for, which corresponds to "true" in the errorExpected field).
> ```

nit: adjust comment about "true"


<!-- Sent from Reviewable.io -->

jackcwu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 45 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRWMn7c7_gzbYe2Yd:-MrTrixzBa9Y4u3eR3xF:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L45)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you can use `require.EqualError` instead of this manual check
</blockquote></details>

fixed
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 63 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRaYZCG2dsqrOGwBR:-MrTrk7gCVIX4LqtCaPn:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L63)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: use `require.EqualError`
</blockquote></details>

fixed
___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 143 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrTRfM92j62FjaEpg-Z:-MrTrlGUAnPN_ePVACyY:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f958ff0621f85712d2081b043e2d62437adb8ca/pkg/cmd/roachtest/tests/validate_grant_option.go#L143)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: adjust comment about "true"
</blockquote></details>

fixed
___
*[pkg/sql/catalog/descriptor.go, line 69 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3h-r6-69:-MqtoNG99cPw4atNjcqg:brythih) ([raw file](https://github.com/cockroachdb/cockroach/blob/53cc4800d7840c99ce8bb604dd15bbaecb786067/pkg/sql/catalog/descriptor.go#L69)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I'd rather call this `MigrationWorkFunc`
</blockquote></details>

renamed
___
*[pkg/sql/catalog/descriptor.go, line 92 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3h-r6-92:-Mqtoak59fsj-fHYRQ3U:b-hff6wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/53cc4800d7840c99ce8bb604dd15bbaecb786067/pkg/sql/catalog/descriptor.go#L92)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Prefer `RunMigrationFuncs` or `RunMigrationChanges`
</blockquote></details>

renamed
___
*[pkg/sql/catalog/dbdesc/database_desc_builder.go, line 79 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3g-r6-79:-Mqtq31P2WtYq712t3At:bpxz1u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/53cc4800d7840c99ce8bb604dd15bbaecb786067/pkg/sql/catalog/dbdesc/database_desc_builder.go#L79)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

s/workFunc/workFuncs
</blockquote></details>

renamed to `workFuncs`


<!-- Sent from Reviewable.io -->

postamar reviewed (changes requested):
I have a few changes to propose, which should be minor.

One thing I'm still not clear on is how do we expect to deal with backups which are restored after this migration runs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-:-MrYb7ml7nL-sjBpSVKK:bkr5oy6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jackcwu, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYYG_iC7lnc4rsY3kK:-MrYYG_j083_5i7E2anJ:bgl2vgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L67)):*
> ```Go
> 			}
> 
> 			err := b.RunMigrationChanges(maybeAddGrantOptions)
> ```

Okay, I finally get why you wanted this `RunMigrationChanges` thing which, until now, hadn't made much sense to me. It's not the worst thing in the world but I still believe that you don't actually need it.

For a start, I do believe that you still need to run `RunPostDeserializationChanges` anyway because I'm almost certain that you'll want to have `catprivilege.MaybeFixPrivileges` run before running `catprivilege.MaybeUpdateGrantOptions `. This way you're sure that the descriptor actually has a privilege descriptor in the first place.

Next, nothing says that you can _only_ change your descriptor in the builder, you can instead do this:
```
b := catalogkv.NewBuilderWithMVCCTimestamp(&descs[i], timestamps[i])
if b == nil {
    return errors.Newf("unable to find descriptor for id %d", id)
}

if err := b.RunPostDeserializationChanges(ctx, nil /* dg */ ); err != nil {
    return err
}

mut := b.BuildExistingMutable()

// Update GRANT options, if required.
switch mut := mut.(type) {
case *dbdesc.DatabaseDescriptor:
        catprivilege.MaybeUpdateGrantOptions(&mut.Privileges)
    case *schemadesc.SchemaDescriptor:
        catprivilege.MaybeUpdateGrantOptions(&mut.Privileges)
    ...
}


modifiedDescs = append(modifiedDescs, mut)

```

Perhaps there's something I'm still not seeing.
___
*[pkg/migration/migrations/grant_option_migration.go, line 149 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYWt8V53EmjaWR13wj:-MrYWt8V53EmjaWR13wk:bf02ox7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L149)):*
> ```Go
> 		batch := txn.NewBatch()
> 		for _, desc := range modifiedDescs {
> 			err := catalogkv.WriteDescToBatch(ctx, false, d.Settings, batch, d.Codec, desc.GetID(), desc)
> ```

I believe you should be calling `descriptors.WriteDescToBatch` here instead, otherwise the descriptor version counter won't be incremented, which is bad. 

Calls to `catalogkv` should be avoided, they're typically not what anyone wants, instead there's usually a `descs.Collection` equivalent. We on the Schema team fully realize that none of this is obvious and we _do_ intend to do something about it at some point.
___
*[pkg/migration/migrations/grant_option_migration_external_test.go, line 191 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYX5xP5k9V2mRo7qXz:-MrYX5xP5k9V2mRo7qY-:bcok2wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration_external_test.go#L191)):*
> ```Go
> 			// GetAllDescriptors calls PrivilegeDescriptor.Validate, all the invalid
> 			// descriptors should be updated.
> 			descs, err := catalogkv.GetAllDescriptors(ctx, txn, keys.SystemSQLCodec, false /* shouldRunPostDeserializationChanges */)
> ```

Use the `descs.Collection` instead of `catalogkv` calls whenever possible. This is a test, so it's not a huge deal, but still.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jackcwu, @postamar, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYYG_iC7lnc4rsY3kK:-MrYtUsG2Zgf5NPAbjbZ:b-qld49x) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L67)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Okay, I finally get why you wanted this `RunMigrationChanges` thing which, until now, hadn't made much sense to me. It's not the worst thing in the world but I still believe that you don't actually need it.

For a start, I do believe that you still need to run `RunPostDeserializationChanges` anyway because I'm almost certain that you'll want to have `catprivilege.MaybeFixPrivileges` run before running `catprivilege.MaybeUpdateGrantOptions `. This way you're sure that the descriptor actually has a privilege descriptor in the first place.

Next, nothing says that you can _only_ change your descriptor in the builder, you can instead do this:
```
b := catalogkv.NewBuilderWithMVCCTimestamp(&descs[i], timestamps[i])
if b == nil {
    return errors.Newf("unable to find descriptor for id %d", id)
}

if err := b.RunPostDeserializationChanges(ctx, nil /* dg */ ); err != nil {
    return err
}

mut := b.BuildExistingMutable()

// Update GRANT options, if required.
switch mut := mut.(type) {
case *dbdesc.DatabaseDescriptor:
        catprivilege.MaybeUpdateGrantOptions(&mut.Privileges)
    case *schemadesc.SchemaDescriptor:
        catprivilege.MaybeUpdateGrantOptions(&mut.Privileges)
    ...
}


modifiedDescs = append(modifiedDescs, mut)

```

Perhaps there's something I'm still not seeing.
</blockquote></details>

My preferences is that we move the upgrading of the privileges into the `PostDeserializationChanges`. We can then plumb whatever boolean or version we need into that method. That way we'll be able to comfortably retain this logic in our existing framework that will apply equally to restores and what not. Consider: 

```go
-       RunPostDeserializationChanges(ctx context.Context, dg DescGetter) error
+       RunPostDeserializationChanges(
+               ctx context.Context, dg DescGetter, version *clusterversion.ClusterVersion,
+       ) error
```

If we did that, and then we did two versions, one to activate the new behavior and a subsequent one to upgrade the representation, I think we'd be in a good place. Thoughts?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jackcwu, @postamar, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYYG_iC7lnc4rsY3kK:-MrZwXB92qgx7NU5wfh2:b-3fin3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L67)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

My preferences is that we move the upgrading of the privileges into the `PostDeserializationChanges`. We can then plumb whatever boolean or version we need into that method. That way we'll be able to comfortably retain this logic in our existing framework that will apply equally to restores and what not. Consider: 

```go
-       RunPostDeserializationChanges(ctx context.Context, dg DescGetter) error
+       RunPostDeserializationChanges(
+               ctx context.Context, dg DescGetter, version *clusterversion.ClusterVersion,
+       ) error
```

If we did that, and then we did two versions, one to activate the new behavior and a subsequent one to upgrade the representation, I think we'd be in a good place. Thoughts?
</blockquote></details>

Actually, I don't think we need to plumb a version down. It seems to me that one thing we can do is to have post deserialization changes by default in this version always set the relevant grant options given the grant privilege, but leave the grant privilege in place. It's an additive and backwards compatible thing to set, right? We won't be able to upgrade the version at this point. Then, we can have some mechanism to tell the builder that in post deserialization changes it should actually remove that grant privilege. We only need to make that call to tell it to do the removal when doing restores and when doing the migration. Then, in the following version, we can just always remove the grant privilege unconditionally in post-deserialization changes and upgrade the version every time. 


<!-- Sent from Reviewable.io -->

jackcwu commented:
Hey everyone, it's my last day so unfortunately I won't be able to complete the suggested changes. Hopefully someone from SQL Experience can take over. I have some notes on additional things that need to be done:

- I will be merging the GRANT privilege deprecation PR (https://github.com/cockroachdb/cockroach/pull/74210) prior to this, so the mixed cluster version test (`pkg/cmd/roachtest/tests/validate_grant_option.go`) and migration test (`pkg/migration/migrations/grant_option_migration_external_test.go`) will need to be modified to accomodate the new changes (basically, whenever GRANT or ALL is granted to someone, they will get grant options on every privilege they have once everything is updated to 22.1. Previously, this was not the case. This behavior should be removed once the GRANT privilege is removed in 22.2)

rafiss reviewed (commented):
> Dismissed @RichardJCai from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jackcwu, @postamar, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/cmd/roachtest/tests/validate_grant_option.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3l-r3-50:-Msmtm-e5K0F82tFX5f8:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/cmd/roachtest/tests/validate_grant_option.go#L50)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

fixed
</blockquote></details>

done
___
*[pkg/migration/migrations/grant_option_migration.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3m-r3-30:-MsmtZkx3XoLc5joZ9Fy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/migration/migrations/grant_option_migration.go#L30)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

added
</blockquote></details>

done
___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYYG_iC7lnc4rsY3kK:-Msmt108-9dafRCuhjg5:bqt5jwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L67)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Actually, I don't think we need to plumb a version down. It seems to me that one thing we can do is to have post deserialization changes by default in this version always set the relevant grant options given the grant privilege, but leave the grant privilege in place. It's an additive and backwards compatible thing to set, right? We won't be able to upgrade the version at this point. Then, we can have some mechanism to tell the builder that in post deserialization changes it should actually remove that grant privilege. We only need to make that call to tell it to do the removal when doing restores and when doing the migration. Then, in the following version, we can just always remove the grant privilege unconditionally in post-deserialization changes and upgrade the version every time. 
</blockquote></details>

i moved the logic into RunPostDeserializationChanges. I agree it should work, since grant options are additive and backward-compatible. I don't think I followed the second half of your suggestion, for how we can remove the old GRANT privilege. Let me know what else is needed.
___
*[pkg/migration/migrations/grant_option_migration.go, line 149 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYWt8V53EmjaWR13wj:-MsmmwOB8AzZ-HfEO9n6:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L149)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I believe you should be calling `descriptors.WriteDescToBatch` here instead, otherwise the descriptor version counter won't be incremented, which is bad. 

Calls to `catalogkv` should be avoided, they're typically not what anyone wants, instead there's usually a `descs.Collection` equivalent. We on the Schema team fully realize that none of this is obvious and we _do_ intend to do something about it at some point.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> Dismissed @RichardJCai from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar, @rafiss, @RichardJCai, and @stevendanna)

___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3j-r3-119:-MsmtvInDzep-htxwopr:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/sql/catalog/tabledesc/table_desc_builder.go#L119)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

changed to call that in the function directly.

I realized that I accidentally left in `tdb.changes, err = maybeFillInDescriptor(ctx, dg, tdb.maybeModified, tdb.skipFKsWithNoMatchingTable)`, which is something called in `runPostDesserializationChanges`. I believe this means that the error can always be nil for this function.
</blockquote></details>

done
___
*[pkg/migration/migrations/grant_option_migration_external_test.go, line 191 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYX5xP5k9V2mRo7qXz:-Msmu-1g0rXSIGJVAotj:b3h6vcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration_external_test.go#L191)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Use the `descs.Collection` instead of `catalogkv` calls whenever possible. This is a test, so it's not a huge deal, but still.
</blockquote></details>

do you mean I should use `tc.Server(0).CollectionFactory` here?
___
*[pkg/cmd/roachtest/tests/privilege_version_upgrade.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-Mosz3NlysLTwz1yWK3d-r3-50:-Msmv7Py976TouZBuKiO:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8df8d58cc4e30f4f6a7622ff67e6f3d2442ba99/pkg/cmd/roachtest/tests/privilege_version_upgrade.go#L50)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

removed
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss commented:
@RichardJCai @ajwerner @postamar this is ready for another review

postamar reviewed (approved):
Sorry this slipped under my radar. This all looks good enough to me. Caveat: I don't know much about the behaviour expected of privileges.

If it's feasible, consider adding a test case in `backupccl` in which we check that restoring a backup from several versions ago produces correct privilege descriptors. I don't insist on it, but it would be a nice to-have, if it's easy to do.

> Reviewed 1 of 7 files at r7, 1 of 5 files at r8, 3 of 6 files at r11, 9 of 13 files at r12, 7 of 7 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-:-MtsFeam2aiyyS2oFN4w:b-vs4zti)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jackcwu, @rafiss, and @RichardJCai)

___
*[pkg/migration/migrations/grant_option_migration.go, line 67 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYYG_iC7lnc4rsY3kK:-MtsF7d_8dB2R5yd2znf:b-tl7r5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration.go#L67)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i moved the logic into RunPostDeserializationChanges. I agree it should work, since grant options are additive and backward-compatible. I don't think I followed the second half of your suggestion, for how we can remove the old GRANT privilege. Let me know what else is needed.
</blockquote></details>

FYI, I don't know enough about privileges to comment further in this thread.
___
*[pkg/sql/catalog/catprivilege/fix.go, line 145 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MtsDUvM4M5vSRf2P0La:-MtsDUvM4M5vSRf2P0Lb:b-y3onwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2248306ba53d320963338fb59ae3a843dc3e07f/pkg/sql/catalog/catprivilege/fix.go#L145)):*
> ```Go
> // if they have the GRANT privilege - if so, then set the user's grant option
> // bits equal to the privilege bits.
> func MaybeUpdateGrantOptions(ptr **descpb.PrivilegeDescriptor) bool {
> ```

nit: This doesn't need to be a pointer to a pointer. We had to do this for MaybeFixPrivileges in case the privileges pointer was `nil` (in which case we make it point to a new empty struct) but when this function is called it's guaranteed to not be nil anymore.
___
*[pkg/migration/migrations/grant_option_migration_external_test.go, line 191 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYX5xP5k9V2mRo7qXz:-MtsEelv3t3Bg-Iv0pi6:bjlnp6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration_external_test.go#L191)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you mean I should use `tc.Server(0).CollectionFactory` here?
</blockquote></details>

Yes, although you might find it simpler to open a transaction with `sql.DescsTxn` which requires the `ExecutorConfig`. It's the same thing under the hood.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/sql/catalog/catprivilege/fix.go, line 145 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MtsDUvM4M5vSRf2P0La:-MttStFa34mOfAYNSL2b:bqzkdtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2248306ba53d320963338fb59ae3a843dc3e07f/pkg/sql/catalog/catprivilege/fix.go#L145)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: This doesn't need to be a pointer to a pointer. We had to do this for MaybeFixPrivileges in case the privileges pointer was `nil` (in which case we make it point to a new empty struct) but when this function is called it's guaranteed to not be nil anymore.
</blockquote></details>

ah ok, fixed
___
*[pkg/migration/migrations/grant_option_migration_external_test.go, line 191 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYX5xP5k9V2mRo7qXz:-MttRxvBEU4ycu3uJuZq:blejzh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration_external_test.go#L191)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Yes, although you might find it simpler to open a transaction with `sql.DescsTxn` which requires the `ExecutorConfig`. It's the same thing under the hood.
</blockquote></details>

hm, actually, i need to keep using the `catalogkv` version here, since i need shouldRunPostDeserializationChanges=false


<!-- Sent from Reviewable.io -->

rafiss committed:
roachtest,sql: enhance grant option migration test and fix bugs

This changes the migration so that it runs on all the privilege
descriptors that don't have grant options set.

Additionally, the GRANT/REVOKE logic is changed so that it also applies
the grant option if the grantee currently has the GRANT privilege.

Release note: None

postamar reviewed (commented):
> Reviewed 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[pkg/migration/migrations/grant_option_migration_external_test.go, line 191 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/72982#-MrYX5xP5k9V2mRo7qXz:-MtxkHU83hYutr1h4zPa:b-sc1lzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677497c7bd841fbe6fe4565ee35e75071e2c63a/pkg/migration/migrations/grant_option_migration_external_test.go#L191)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, actually, i need to keep using the `catalogkv` version here, since i need shouldRunPostDeserializationChanges=false
</blockquote></details>

Oh, good point. Sorry for misleading you.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/migration/migrations/grant_option_migration.go:
```diff
@@ -44,6 +44,23 @@ func grantOptionMigration(
 	addGrantOptionFunc := func(ids []descpb.ID, descs []descpb.Descriptor, timestamps []hlc.Timestamp) error {
 		var modifiedDescs []catalog.MutableDescriptor
 		for i, id := range ids {
+			// Temporarily change the version in order to force the migration to run
```
> Doesn't this make it so the migration can run twice - ie we don't have a base case for the descriptor state to reach?

Sort of -- but the migration itself only runs when the clusterversion reaches `clusterversion.ValidateGrantOption` and that can only happen once.

> Also are we sure the grant option bits are wiped out?
Yes, this was causing line 178 to fail here in validate_grant_option.go  ( https://github.com/cockroachdb/cockroach/pull/72982/commits/aafdb46c32751e956cf4180e3bef1b8cdffd7fe9#diff-edd6b676708815ac3b09d3f07bb9c8c4bffe7441dacc9452c6f7453c2ba194efR178 )

rafiss replied:
here's the reason it gets wiped out:

- suppose you have a mixed-version cluster
- run `CREATE TABLE t` on a node running the new binary. this creates a privilege descriptor with a version of `GrantOptionVersion`.
- run `GRANT GRANT, SELECT ON t TO testuser` on a node running the new binary. `testuser` can now grant SELECT to others, and the descriptor has its grant options bits set.
- run `GRANT DELETE ON t TO testuser` on a node running the old binary. now the descriptor has the GRANT, SELECT, and DELETE privileges, but the grant option bits are not set since the old binary cannot serialize a descriptor with that info.
- finalize the cluster upgrade. without my change here, the grant option bits are not added, since the privilege descriptor already is at `GrantOptionVersion`. now `testuser` can NOT grant SELECT to others, even though they could right before the upgrade was finalized.

RichardJCai replied:
Ah I see, it's not wiping out the grant bit but the migration isn't happening as expected.

This migration is kind of thorny, I'm still a bit iffy on downgrading the version temporarily, it just seems scary since we generally always expect the version upgrade to increase.

What if in this step
```
run GRANT GRANT, SELECT ON t TO testuser on a node running the new binary. testuser can now grant SELECT to others, and the descriptor has its grant options bits set.
```
We just also lazily upgrade the descriptor to have GRANT bits, and continue keeping GRANT privilege as well as we still need it until we finalize the migration, then in the next version we'll have a migration if necessary to then fully remove GRANT option (we can't do it til then)

rafiss replied:
> it just seems scary since we generally always expect the version upgrade to increase.

well, the version is only modified in memory here; we never write that back to disk.

> We just also lazily upgrade the descriptor to have GRANT bits, and continue keeping GRANT privilege as well as we still need it until we finalize the migration

This is how it works right now.

>  it's not wiping out the grant bit but the migration isn't happening as expected.

It is wiping out the grant option bits

rafiss replied:
I view the temporary, in-memory change to the version as safe. It's just a way to avoid having to do another migration in the 22.1->22.2 upgrade. with the migration like this, in during the 22.2 we should be able to just stop referencing the GRANT privilege in the code without needing to do anything else. (we'd still have to keep the GRANT enum value, so that it never accidentally gets reused in the future.)

RichardJCai replied:
Sorry, I misunderstood. I see now, we moved some of the logic back into `RunPostDeserializedChanges`.

I'm still somewhat confused on how the grant bits get wiped, I thought since on the old node we wouldn't be updating the grant bits in the proto, they'd stay the same. Gonna check out the branch to double check my understanding.

My concern with downgrading the version is that it's kind of strange that we're doing the migration on privilege descriptors that shouldn't need the migration.

Ideally for me, new descriptors should not need the migration at all if they're created on that version.
Can we version gate the descriptor version as well? Only set the descriptor version to `GrantOptionVersion` iff the cluster verison is `ValidateGrantOption`?

rafiss replied:
> I'm still somewhat confused on how the grant bits get wiped, I thought since on the old node we wouldn't be updating the grant bits in the proto, they'd stay the same.

it's because the old node's proto definition doesn't even have a field for the grant option bits. so when it reads the bytes and turns it into a privilege descriptor struct, it doesn't have that field set at all. and when it turns it back into bytes, it wouldn't even know how to serialize that field anyway.

> new descriptors should not need the migration at all if they're created on that version.
Can we version gate the descriptor version as well? Only set the descriptor version to GrantOptionVersion iff the cluster verison is ValidateGrantOption?

yes this would work too. i didn't do it though since it would require a lot of plumbing, which we'd then want to undo in 22.2.

> My concern with downgrading the version is that it's kind of strange that we're doing the migration on privilege descriptors that shouldn't need the migration.

Well the thing is, they do need the migration. If we go with your suggestion of setting GrantOptionVersion iff the cluster verison is ValidateGrantOption, then the migration would work the exact same way, in practice -- it would run on all privilege descriptors.

if it helps, i can do one more check for this "temporary version change" bit i added -- it only needs to run IFF there are no grant option bits currently set.

RichardJCai replied:
Yeah looking into it, the plumbing is pretty tedious. Let me think about it for a little more to see if we can avoid it, otherwise it's fine.

RichardJCai commented on pkg/sql/catalog/catprivilege/fix.go:
```diff
@@ -138,3 +138,36 @@ func MaybeFixPrivileges(
+// bits equal to the privilege bits.
+func MaybeUpdateGrantOptions(p *descpb.PrivilegeDescriptor) bool {
+	changed := false
+	if p.Version >= descpb.GrantOptionVersion {
```
I think if we're decreasing the version, then we don't need this new version at all since this is the only place where the version is used and checked. I had envisioned adding this new version would make us run the migration only on the old descriptors but if we're decreasing the version regardless then it seems like this new version wouldn't be needed.

Perhaps it makes sense to do this check
```
			if p != nil {
				// If root doesn't have these grant option bits set, then we know
				// that an old node must have clobbered the grant option bits.
				if !p.CheckGrantOptions(security.RootUserName(), privilege.List{privilege.ALL}) {
					p.Version = descpb.Version21_2
				}
			}
```

here instead, if root has grant option bits then we can skip the migration?

rafiss replied:
hm yeah, that seems reasonable. the only point i could see is that setting the version makes things very explicit. but i think ultimately since this migration is not backwards incompatible, you're right we don't need a new version at all.

rafiss replied:
actually, i still think the GrantOptionVersion could be valuable for one more reason: for system tables, admin/root do not have ALL privileges or grant options. We don't want to run `MaybeUpdateGrantOptions` each time we read a system table descriptor, so incrementing the version allows us to avoid that.

RichardJCai replied:
We could directly use `IsSystemID` and skip if it is right?

rafiss replied:
no, that doesn't work since the admins should still get the SELECT grant option for the system tables; but i think we should be able to use a different check. admins should have the SELECT grant option on every table no matter what, so that should work

RichardJCai reviewed (approved):
Thanks for taking this over and doing the revisions, much appreciated. LGTM now!

rafiss commented:
tftrs all!

bors r=postamar,RichardJCai

rafiss commented:
bors r=postamar,RichardJCai
